### PR TITLE
Use balena bootloader

### PR DESCRIPTION
**Pull request by alexgg**
Add a second stage boot from U-boot into the balena bootloader. This allows to have a common place to implement a/b rollback and bootloader configuration, and is a requirement for secure boot as it needs to decrypt the root partitions before launching the final kernel.

Depends-on: https://github.com/balena-os/meta-balena/pull/3539
After (rebase over): https://github.com/balena-os/balena-iot-gate-imx8/pull/575

### CODE REVIEW

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+diff --git a/arch/arm64/configs/cl-imx8m-mini_defconfig b/arch/arm64/configs/cl-imx8m-mini_defconfig
+index 8fe17f8c4268..fcfedced9cd5 100644
+--- a/arch/arm64/configs/cl-imx8m-mini_defconfig
++++ b/arch/arm64/configs/cl-imx8m-mini_defconfig
+@@ -891,3 +891,5 @@ CONFIG_XEN_I2C_BACKEND=y
+ CONFIG_XEN_WDT=y
+ CONFIG_XEN=y
+ CONFIG_XFRM_USER=m
++CONFIG_CMDLINE="balena_stage2 nr_cpus=1"
```
Comment (alexgg) [inline layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch:21]: I am reviewing whether the `nr_cpus=1` is still required in this DT.

Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch]: It wasn't required so I removed it.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-compulab_5.15.%.bbappend [hunk 2]
```diff
@@ -24,8 +24,6 @@ SCMVERSION="n"
 BALENA_CONFIGS:append = " imx-sdma "
 BALENA_CONFIGS[imx-sdma] = " \
 	CONFIG_IMX_SDMA=m \
-	CONFIG_EFI_STUB=n \
-	CONFIG_EFI=n \
```
Comment (jakogut) [inline layers/meta-balena-imx8mm/recipes-kernel/linux/linux-compulab_5.15.%.bbappend:28]: What purpose does enabling EFI fulfill on a non-EFI system?

**Review by jakogut — APPROVED** [review]
Looks good to me, @alexgg. Nice find on the patch to enable OCOTP clocks to fix kexec.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-compulab_5.15.%.bbappend (hunk 2 above)
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-compulab_5.15.%.bbappend:28]: Only EFI enabled kernels can be signed and authenticated by kexec @jakogut

**Comment by acostach** [comment]
looks good to me, I suspect rollback tests in the autokit may fail because the board uses the new u-boot already and older images don't ship dtbs nor the balena bootloader in the boot partition. For new DTs based on the iot-gate-imx8 we will probably need to increase the boot partition, just in case the compressed bootloader gets past 20MB at some point.

**Comment by alexgg** [comment]
> looks good to me, I suspect rollback tests in the autokit may fail because the board uses the new u-boot already and older images don't ship dtbs nor the balena bootloader in the boot partition.

The only rollback that doesn't work is altboot, and that is because when kexec does the partition switch the old kernel does not have the ocotp patch to allow it to boot from kexec.

For healthcheck, the rollback hooks re-program the old U-boot which bypasses the balena bootloader.

Possible solutions include:

* Have abroot detect it is switching to a non-balena bootloader partition and reprogram the old U-boot that will look for the right kernel in the rootfs
* Make an initial release that patches the kernel with the ocotp patch, and then only allow HUPs from that release to a balena bootloader enabled one.

> For new DTs based on the iot-gate-imx8 we will probably need to increase the boot partition, just in case the compressed bootloader gets larger with BSP updates

Yes, we probably should. I hope we can also use the takeover tool for the `iot-gate-imx8` if the needs arises.

File: layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/abroot [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    fi
+
+    # Rollback if this is the 3rd boot attempt
+    if [ "${BOOT_COUNT}" -ge 3 ]; then
+        if [ "${ROOT_PART}" = "A" ]; then
+            ROOT_PART="B"
+        else
+            ROOT_PART="A"
+        fi
+        export bootparam_rollback=true
```
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/abroot:94]: hey @mtoman this is part of the last commit which is marked as [REVIEW] - I am just testing the solution at this point, so I will clean up later.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/arm64-kexec_file-use-more-system-keyrings-to-verify-.patch [hunk 4]
```diff
@@ -0,0 +1,65 @@
+From: Coiby Xu <coxu@redhat.com>
+Date: Thu, 14 Jul 2022 21:40:26 +0800
+Subject: [PATCH] arm64: kexec_file: use more system keyrings to verify kernel
+ image signature
+
+Currently, when loading a kernel image via the kexec_file_load() system
+call, arm64 can only use the .builtin_trusted_keys keyring to verify
```
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/arm64-kexec_file-use-more-system-keyrings-to-verify-.patch:7]: @mtoman authentication does not work without them - in any case they come from the 5.15.y branch so they are safe to include.

File: layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/rollback [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        # Re-program the old U-boot to bypass balena bootloader as the old kernel
+        # might not be able to kexec without patches.
+        warn "Rolling back U-boot"
+        echo 0 > /sys/block/${BOOTDEV}/force_ro
+        dd if="${OLDROOT_MOUNT}/resin-boot/${UBOOT_FILE}" of=/dev/${BOOTDEV} conv=fdatasync seek=33 bs=1024
+        echo 1 > /sys/block/${BOOTDEV}/force_ro
+    fi
+    umount "${OLDROOT_MOUNT}"
+    # This will reboot
+    exit 1
```
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/rollback:51]: No particular reason - reboot is a command while exit a built-in, so I instinctively tend to the path that requires less extras. I will probably change to reboot once I finish testing it.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+diff --git a/arch/arm64/configs/cl-imx8m-mini_defconfig b/arch/arm64/configs/cl-imx8m-mini_defconfig
+index 8fe17f8c4268..cd8c88e97091 100644
+--- a/arch/arm64/configs/cl-imx8m-mini_defconfig
++++ b/arch/arm64/configs/cl-imx8m-mini_defconfig
+@@ -891,3 +891,5 @@ CONFIG_XEN_I2C_BACKEND=y
+ CONFIG_XEN_WDT=y
+ CONFIG_XEN=y
+ CONFIG_XFRM_USER=m
++CONFIG_CMDLINE="balena_stage2"
```
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch:22]: I will review this - the thinking behind hardcoding it in the kernel was not to depend on U-boot as the iot-gate-imx8 cannot boot from external media, it needs to boot from an internal U-Boot.

Thinking through it again, the old internal U-boot booting flasher will boot the rootfs kernel and not the balena bootloader, so passing this as an environment variable in U-Boot probably works for all three use cases:
* Flasher from older release
* Booting a new release
* Flashing a locked device

File: layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/abroot (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/abroot]: As discussed, this is now a custom abroot script on the meta-balena-imx8mm layer.

File: layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/rollback (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-core/initrdscripts/files/rollback]: This file has been removed.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/arm64-kexec_file-use-more-system-keyrings-to-verify-.patch (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/arm64-kexec_file-use-more-system-keyrings-to-verify-.patch:7]: As discussed, this is actually required.

File: layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexgg) [reply layers/meta-balena-imx8mm/recipes-kernel/linux/linux-balena-bootloader/defconfig-cl-imx8m-mini-extend-bootloader-command-li.patch:22]: As discussed, I have now removed the kernel append and added it as a bootloader configuration as in other platforms.

**Comment by mtoman** [comment]
lgtm

**Review by mtoman — APPROVED** [review]

**Pull request merged**